draft-ietf-jsonpath-base.md

```diff
@@ -1411,82 +1410,101 @@ a type system is first introduced.
-| `Nothing`         | Singular Path or `filter-path` resulting in an empty nodelist      |
-| `Nodes(nl)`       | `filter-path` resulting in the non-empty nodelist `nl`             |
+| `Nothing`         | A representation of the absence of a JSON value, distinct from the JSON literal `null` |
+| `Nodes(nl)`       | `filter-path` resulting in the nodelist `nl`, which may or may not be empty  |
 {: #tbl-typerep title="Concrete representations of abstract instances"}
 
-The following subtype relationships depend on coercion:
+### Type Conversions {#type-conv}
+
+The following type conversions may occur:
 
-* `OptionalNodeType` is a subtype of `OptionalValueType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalValueType` instance `Value(v)`, where `v` is the value of the node `n`.
-* `OptionalNodeType` is a subtype of `OptionalNodesType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalNodesType` instance `Nodes(l)`, where `l` is a nodelist consisting of just the node `n`.
+* A member of `NodelistType` may be converted to a `ValueType` as follows:
+  * If the nodelist contains a single node, the conversion result is the value of the node.
+  * If the nodelist is empty or contains multiple nodes, the conversion result is `Nothing`.
+* A member of `NodelistType` may be converted to a `BooleanType` as follows:
+  * If the nodelist contains one or more nodes, the conversion result is `BooleanTrue`.
+  * If the nodelist is empty, the conversion result is `BooleanFalse`.
 
 The type correctness of function expressions can now be defined in terms of this type system.
 
 ### Type Correctness of Function Expressions
 
-A function expression is correctly typed if all the following are true:
-
-* If it occurs directly in a test expression, the function
-is defined to have a result type in `ST(OptionalNodesType)`.
-or to have a result type in `ST(OptionalBooleanType)`.
-* If it occurs as a `comparable` in a comparison, the function
-is defined to have a result type in `ST(OptionalNodeOrValueType)`.
-* For it and any function expression it contains,
-each argument of the function matches the defined type of the argument
-according to one of the following rules:
-  * The argument is a function expression with defined result type
-    that is the same as, or a subtype of, the defined type of the argument.
-  * The argument is a literal primitive value and the defined type of the  argument is `Value` or any type of which `Value` is a subtype.
-  * The argument is a Singular Path and the defined type of the argument is `OptionalNodeType` or any type of which `OptionalNodeType` is a subtype.
-  * The argument is a `filter-path` (which includes Singular Paths) and the defined type of the argument is `OptionalNodesType`.
+A function expression is typed based on its return, and may only appear in expressions as follows:
```

Comment:
There are now no typing rules for functions calling other functions. Surely the result type of a function used as a function argument should line up with the argument type.

Reply:
> Surely the result type of a function used as a function argument should line up with the argument type.

Yes.  I think this is implied by typing the function.  If function `foo()` is used as an argument for function `bar()`, i.e. `bar(foo())`, then `foo()` MUST return the same type that `bar()` expects to receive.

We can explicitly state this if you like, but I'm not sure it's necessary to.